bip-0324.mediawiki

```diff
@@ -0,0 +1,587 @@
+
+Encryption and decryption of packets then follow by composing the ciphers from the previous section as building blocks.
+
+<pre>
+LENGTH_FIELD_LEN = 3
+HEADER_LEN = 1
+IGNORE_BIT_POS = 7
+
+def v2_enc_packet(peer, contents, aad=b'', ignore=False):
+    assert len(contents) <= 2**24 - 1
+    header = (ignore << IGNORE_BIT_POS).to_bytes(HEADER_LEN, 'little')
+    plaintext = header + contents
+    aead_ciphertext = peer.send_P.encrypt(aad, plaintext)
+    enc_contents_len = peer.send_L.encrypt(len(contents).to_bytes(LENGTH_FIELD_LEN, 'little'))
+    return enc_contents_len + aead_ciphertext
+</pre>
+
+<pre>
+CHACHA20POLY1305_EXPANSION = 16
+
+def v2_receive_packet(peer, aad=b'', skip_decoy=True):
+    while True:
+        enc_contents_len = receive(peer, LENGTH_FIELD_LEN)
+        contents_len = int.from_bytes(peer.recv_L.crypt(enc_contents_len), 'little')
+        aead_ciphertext = receive(peer, HEADER_LEN + contents_len + CHACHA20POLY1305_EXPANSION)
+        plaintext = peer.recv_P.decrypt(aead_ciphertext)
+        if plaintext is None:
+            disconnect(peer)
+            break
+        header = plaintext[:HEADER_LEN]
+        if not (skip_decoy and header[0] & (1 << IGNORE_BIT_POS)):
+            return plaintext[HEADER_LEN:]
+</pre>
+
+==== Performance ====
+
+Each v1 P2P message uses a double-SHA256 checksum truncated to 4 bytes. Roughly the same amount of computation power is required for encrypting and authenticating a v2 P2P message as proposed.
+
+=== Application layer specification ===
+==== v2 Bitcoin P2P message structure ====
+v2 Bitcoin P2P transport layer packets use the encrypted message structure shown above. An unencrypted application layer '''contents''' is composed of:
+
+{|class="wikitable"
+! Field !! Size in bytes !! Comments
+|-
+| <code>message_type</code> || ''1..13'' || either a one byte ID or an ASCII string prefixed with a length byte
+|-
+| <code>message_payload</code> || <code>message_length</code> || message payload
+|}
+
+If the first byte of <code>message_type</code> is in the range ''1..12'', it is interpreted as the number of ASCII bytes that follow for the message type. If it is in the range ''13..255'', it is interpreted as a message type ID. This structure results in smaller messages than the v1 protocol as most messages sent/received will have a message type ID.<ref name="smaller_messages">'''How do the length between v1 and v2 compare?''' For messages that use the 1-byte short message type ID, v2 packets use 3 bytes less per message than v1.</ref>
+
+The following table lists currently defined message type IDs:
+
+{| class="wikitable"
+|-
+!
+!0
+!1
+!2
+!3
+|-
+!+0
+|(undefined)||(1 byte string)||(2 byte string)||(3 byte string)
+|-
+!+4
+|(4 byte string)||(5 byte string)||(6 byte string)||(7 byte string)
+|-
+!+8
+|(8 byte string)||(9 byte string)||(10 byte string)||(11 byte string)
+|-
+!+12
+|(12 byte string)||<code>ADDR</code>||<code>BLOCK</code>||<code>BLOCKTXN</code>
```

Comment:
Is there any reason to expect clients supporting this BIP to use `ADDR` rather than `ADDRV2`?